src/block/block_compact.c

```diff
@@ -65,6 +70,295 @@ __wt_block_compact_get_progress_stats(WT_SESSION_IMPL *session, WT_BM *bm,
     *pages_reviewedp = block->compact_pages_reviewed;
     *pages_skippedp = block->compact_pages_skipped;
     *pages_rewrittenp = block->compact_pages_rewritten;
+    *pages_rewritten_expectedp = block->compact_pages_rewritten_expected;
+}
+
+/*
+ * __block_compact_trim_extent --
+ *     Trim the extent to the given range mask, specified via start and end offsets.
+ */
+static inline void
+__block_compact_trim_extent(WT_SESSION_IMPL *session, wt_off_t mask_start, wt_off_t mask_end,
+  wt_off_t *ext_startp, wt_off_t *ext_sizep)
+{
+    WT_UNUSED(session);
+
+    if (mask_end >= 0 && mask_end < mask_start) {
+        *ext_sizep = 0;
+        return;
+    }
+
+    /* Trim from the beginning. */
+    if (*ext_startp < mask_start) {
+        if (*ext_startp + *ext_sizep <= mask_start) {
+            *ext_sizep = 0;
+            return;
+        }
+        *ext_sizep -= mask_start - *ext_startp;
+        *ext_startp = mask_start;
+    }
+
+    /* Trim from the end. */
+    if (mask_end >= 0 && *ext_startp + *ext_sizep > mask_end) {
+        *ext_sizep = mask_end - *ext_startp;
+        if (*ext_sizep <= 0) {
+            *ext_sizep = 0;
+        }
+    }
+}
+
+/*
+ * __block_compact_skip_internal --
+ *     Return if compaction will shrink the file. This function takes a few extra parameters, so
+ *     that it can be useful for both making the actual compaction decisions as well as for
+ *     estimating work ahead of the compaction itself.
+ */
+static void
+__block_compact_skip_internal(WT_SESSION_IMPL *session, WT_BLOCK *block, bool estimate,
+  wt_off_t file_size, wt_off_t start_offset, wt_off_t avail_bytes_before_start_offset, bool *skipp,
+  int *compact_pct_tenths_p)
+{
+    WT_EXT *ext;
+    wt_off_t avail_eighty, avail_ninety, off, size, eighty, ninety;
+
+    /* IMPORTANT: We assume here that block->live_lock is locked. */
+
+    /* Sum the available bytes in the initial 80% and 90% of the file. */
+    avail_eighty = avail_ninety = avail_bytes_before_start_offset;
+    ninety = file_size - file_size / 10;
+    eighty = file_size - ((file_size / 10) * 2);
+
```

Comment:
FWIW, we have a project schedule to start in the next quarter or so that should address this. The project is basically a set of usability improvements for compact. One of the goals is to allow the user/application to specify the compaction target instead of hard-coding it to 10%. So this code should be changing soon...

The goals for that project are in the scope document linked from [WRITING-12134](https://jira.mongodb.org/browse/WRITING-12134).

Reply:
That's good to hear :) In that case, I'll not act on this, as there are already plans to fix it in a separate project.